### PR TITLE
Transfer images

### PR DESCRIPTION
**Pull request by iameskild**
As we transition to the nebari-dev org, we have now decided to split out the docker images from the old monorepo. 

Fixes: https://github.com/Quansight/qhub/issues/1174
--- 
OUTDATED

A few outstanding items or questions:
- Can I have my permissions for this repo elevated so I can add the appropriate secrets?
- I tried to update the `image.yaml` workflow to only push to GHCR and Quay but it seems that the `docker/build-push-action` will always try to push to Dockerhub. [GHA details in my fork](https://github.com/iameskild/docker-images/runs/6662639852?check_suite_focus=true#step:11:5).
  - This means we have to either find a replacement for the `docker/build-push-action` workflow or registry a Dockerhub account for nebari. Thoughts?

### CODE REVIEW

**Comment by iameskild** [comment]
@costrouc @trallard @magsol this PR is ready for review. I tested it on my forked repo and the images were successfully built and pushed.

**Review by magsol — APPROVED** [review]
LGTM. May still want to wait for an additional review but they build successfully for me.

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:14]: using `**` basically means that changes to any file in this repo this + lines 15-18 seem making each other redundant

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:27]: Can you provide moe context on why we need write access to these two?

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:40]:
```suggestion
      - name: 'Checkout Repository'
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:41]:
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:44]:
```suggestion
        uses: docker/setup-buildx-action@v2
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:46]:
```suggestion
      - name: Login to GitHub Container Registry
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:48]:
```suggestion
        uses: docker/login-action@v2
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:56]:
```suggestion
        uses: docker/login-action@v2
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:63]: TODO: need to add a hadolint.yaml file

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:30]:
```suggestion
  build-images:
```
best to use descriptive names

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:91]:
```suggestion
        uses: docker/build-push-action@v3
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:90]:
```suggestion
      - name: Build docker images
```

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:107]: We are already using the gha cache on  the step above - so this is redundant as well

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:118]: Now - I get why we are doing the cache step not entirely sure on the value but adding a ping here

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:101]: All steps have a name but this one 
TODO: add a meaningful step name

File: README.md
Comment (trallard) [inline README.md:1]: This needs to be also a meaningful name 

```suggestion
# nebari base Docker images
```
 or something similar

File: README.md
Comment (trallard) [inline README.md:4]: This readme went straight to the pinning... as an external contributor, I want to know the following:

1. The CI status - TODO: add badge for the CI
1. What is in this repo - i.e. the purpose of the images
2. How do I get started? what dependencies do I need? 
3. Step by step instructions to build the images because more than likely I will copy and paste in whichever order the instructions come in
4. Can I use them locally? if so how do I do this?
5. What is the license?
6. How do I report issues?
7. Where are the images pushed?

Inspiration:
- https://github.com/jupyterhub/repo2docker
- https://github.com/openshift/source-to-image
- https://github.com/jupyter/docker-stacks

File: README.md
Comment (trallard) [inline README.md:2]: Needs replacing see below

File: jupyterhub/postBuild
Comment (trallard) [inline jupyterhub/postBuild:1]: the fact that it is only a `set....` command makes me wonder if this is necessary

File: scripts/install-conda.sh
Comment (trallard) [inline scripts/install-conda.sh:4]:
```suggestion
wget --quiet -O miniconda.sh https://repo.anaconda.com/miniconda/Miniconda3-$CONDA_VERSION-Linux-x86_64.sh
```

File: scripts/install-conda.sh
Comment (trallard) [inline scripts/install-conda.sh:6]:
```suggestion
echo "${CONDA_SHA256} miniconda.sh" > miniconda.checksum
```

File: scripts/install-conda.sh
Comment (trallard) [inline scripts/install-conda.sh:13]:
```suggestion
```

File: scripts/install-conda.sh
Comment (trallard) [inline scripts/install-conda.sh:23]: 🤔  - instead of installing miniforge then installing mamba would it not make sense to install mambaforge then remove these additional steps?

**Review by trallard — CHANGES_REQUESTED** [review]

**Comment by trallard** [comment]
@iameskild, thanks for opening this PR. I just went through an in-depth review and left several comments. 

Some additional thoughts:
1. Right now the Dockerfiles are in `.` but there are directories for each image - this pattern is counter-intuitive I'd suggest moving the Docker images to the respective directories to keep everything aligned
2. Additionally - and moving forward it be best to give more details on the infrastructure/ config of the repos this can be done either:
	1. Adding a Config files doc (.md) like the [repo2docker one](https://repo2docker.readthedocs.io/en/latest/config_files.html)
	2. Adopting an `INFRASTRUCTURE.md` like the [sphinx-book-theme one](https://github.com/executablebooks/sphinx-book-theme/blob/master/ARCHITECTURE.md), [pymc](https://github.com/pymc-devs/pymc/blob/main/ARCHITECTURE.md)
	
Let me know if there is anything that needs clarification or else 👋🏽  

## Resources
- [A good read on ARCHITECTURE.md files](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html)

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:69]:
```suggestion
      - name: Generate Docker images tags
```

**Comment by trallard** [comment]
> * Can I have my permissions for this repo elevated so I can add the appropriate secrets?
I believe this has been done? - also, when adding secrets can we create an env such as `docker-dev` then place the secrets there? 

> * I tried to update the `image.yaml` workflow to only push to GHCR and Quay but it seems that the `docker/build-push-action` will always try to push to Dockerhub. [GHA details in my fork](https://github.com/iameskild/docker-images/runs/6662639852?check_suite_focus=true#step:11:5).
I went through the logs, and it seems the error is `Error: buildx failed with: error: failed to solve: server message: insufficient_scope: authorization failed`
so that makes me believe you need to change the scope of the PAT

File: .github/workflows/image.yaml
Comment (trallard) [inline .github/workflows/image.yaml:21]: more thoughts: we only want to build and push on creating + publishing a new tag right?

File: .github/workflows/image.yaml
Comment (trallard) [reply .github/workflows/image.yaml:27]: nevermind the security one - it is for code scanner

File: .github/workflows/image.yaml
Comment (trallard) [reply .github/workflows/image.yaml:10]: the correct pattern should be IMHO: 
- PRs -> build do not push
- merge into main -> build and push
- tag/release published -> build and push

as for paths changes triggering the workflow:
- Dockerfile 
- any environment.yml changes
- changes to this file itself


Also can this file be renamed to `build-push-docker.yaml`

**Review by trallard — CHANGES_REQUESTED** [review]

File: .github/workflows/image.yaml
Comment (iameskild) [reply .github/workflows/image.yaml:10]: Done x3 :)

File: .github/workflows/image.yaml
Comment (iameskild) [reply .github/workflows/image.yaml:21]: I could see a scenario where you might want to build and push images that we are testing or experimenting with so something like this might work:
```yaml
on:
  push:
    branches:
      - main
    paths:
      - "Dockerfile.*"
      - "*/environment.yaml"
      - ".github/workflows/image.yaml"
    tags:
      - '*'
```

👆 this is based one of the first comments in this review

File: .github/workflows/image.yaml
Comment (iameskild) [reply .github/workflows/image.yaml:63]: Added.

File: .github/workflows/image.yaml
Comment (iameskild) [reply .github/workflows/image.yaml:107]: @HarshCasper would you mind chiming in here? Are we enabling the Trivy results to be cached or something else?

File: .github/workflows/image.yaml
Comment (iameskild) [reply .github/workflows/image.yaml:101]: Added.

File: README.md
Comment (iameskild) [reply README.md:4]: Added most of these items to the README 👍

File: scripts/install-conda.sh
Comment (iameskild) [reply scripts/install-conda.sh:22]: That makes sense. Miniforge is currently being downloaded from repo.anaconda however we might need to download [mambaforge from github](https://github.com/conda-forge/miniforge#mambaforge).

I will open an issue to track this for now.

**Comment by iameskild** [comment]
Hi @trallard, thanks for the thorough review. Although I tried to incorporate as much of your feedback as possible, I wasn't able to get to all of it. You brought up a lot of good points that are definitely worth discussing with the rest of the team, namely tasks related to https://github.com/Quansight/qhub/issues/1145. This will include these points you raised above:
> Some additional thoughts:
> 
> 1. Right now the Dockerfiles are in `.` but there are directories for each image - this pattern is counter-intuitive I'd suggest moving the Docker images to the respective directories to keep everything aligned
> 2. Additionally - and moving forward it be best to give more details on the infrastructure/ config of the repos this can be done either:
>    
>    1. Adding a Config files doc (.md) like the [repo2docker one](https://repo2docker.readthedocs.io/en/latest/config_files.html)
>    2. Adopting an `INFRASTRUCTURE.md` like the [sphinx-book-theme one](https://github.com/executablebooks/sphinx-book-theme/blob/master/ARCHITECTURE.md), [pymc](https://github.com/pymc-devs/pymc/blob/main/ARCHITECTURE.md)  

Finally, I pinged @HarshCasper to help fill us in on how the Trivy vulnerability scan works but otherwise, I feel like this is ready for a final review :)

**Comment by iameskild** [comment]
Hey @trallard, I believe this PR is ready to be merged, any final thoughts?

File: Dockerfile.dask-worker
Comment (trallard) [inline Dockerfile.dask-worker:9]: TODO: later on - we need to drop 3.7

File: README.md
Comment (trallard) [inline README.md:12]:
```suggestion
- [`nebari-dask-worker`](https://github.com/orgs/nebari-dev/packages/container/package/nebari-dask-worker)
```

File: README.md
Comment (trallard) [inline README.md:24]:
```suggestion
### Pre-requisite
```

keeping Title case

File: README.md
Comment (trallard) [inline README.md:30]:
```suggestion
Assuming you are in the repo's root folder, you can build these images locally by running the listed commands on your terminal.
```

File: README.md
Comment (trallard) [inline README.md:47]:
```suggestion
> **NOTE** 
> It is extremely important to pin specific packages `dask-gateway` and `distributed` as they need to be running the same version for the dask-workers to work as expected.
```

**Review by trallard — APPROVED** [review]
Looks good - added some trivial comments/ suggestions that I will incorporate sight away and merge 🚀

**Pull request merged**
